linera-core/src/worker.rs

```diff
@@ -682,120 +679,85 @@ where
-                    Some(endpoint) => break endpoint,
-                    None => sleep(Duration::from_millis(250)).await,
-                }
-                warn!("No chain worker candidates found for eviction, retrying...");
-            }
-        })
-        .await
-        .map_err(|_| WorkerError::FullChainWorkerCache)?;
-
-        if let Some(receiver) = new_receiver {
+    ) -> Result<Arc<Mutex<ChainWorkerActor<StorageClient>>>, WorkerError> {
+        // If a worker already exists, we only need a read lock.
+        if let Some(worker) = self
+            .chain_workers
+            .read()
+            .await
+            .get(&chain_id)
+            .map(Arc::clone)
+        {
+            return Ok(worker);
+        }
+        // We may have to create a new entry: obtain a write lock.
+        let mut guard = self.chain_workers.write().await;
+        // Between the two locks, another task could have inserted the worker.
+        if let Some(worker) = guard.get(&chain_id).map(Arc::clone) {
+            return Ok(worker);
+        }
+        // No? Then we'll create a new one.
+        let load_worker = async {
             let delivery_notifier = self
                 .delivery_notifiers
                 .lock()
                 .unwrap()
                 .entry(chain_id)
                 .or_default()
                 .clone();
-
-            let actor_task = ChainWorkerActor::run(
+            ChainWorkerActor::load(
                 self.chain_worker_config.clone(),
                 self.storage.clone(),
                 self.block_cache.clone(),
                 self.execution_state_cache.clone(),
                 self.tracked_chains.clone(),
                 delivery_notifier,
                 chain_id,
-                receiver,
-            );
-
-            self.chain_worker_tasks
-                .lock()
-                .unwrap()
-                .spawn_task(actor_task);
-        }
-
-        Ok(sender)
-    }
-
-    /// Retrieves an endpoint to a [`ChainWorkerActor`] from the cache, attempting to create one
-    /// and add it to the cache if needed.
-    ///
-    /// Returns [`None`] if the cache is full and no candidate for eviction was found.
-    #[instrument(level = "trace", skip(self))]
-    #[expect(clippy::type_complexity)]
-    fn try_get_chain_worker_endpoint(
-        &self,
-        chain_id: ChainId,
-    ) -> Option<(
-        ChainActorEndpoint<StorageClient>,
-        Option<
-            mpsc::UnboundedReceiver<(ChainWorkerRequest<StorageClient::Context>, tracing::Span)>,
-        >,
-    )> {
-        let mut chain_workers = self.chain_workers.lock().unwrap();
-
-        if let Some(endpoint) = chain_workers.get(&chain_id) {
-            Some((endpoint.clone(), None))
-        } else {
-            if chain_workers.len() >= usize::from(chain_workers.cap()) {
-                let (chain_to_evict, _) = chain_workers
+            )
+            .await
```

Comment:
This is a little bit subtle and significantly the reason for the complication in the previous initialization code: holding the guard to the map over the await point here leaves us open to deadlocks if the caller doesn't remember to drive this task along with any other task that may access the chain workers map.  It's important that whenever we must hold a lock across an await point, we make sure that any concurrent calls that would try to take the lock also drive the task that is currently holding the lock.